power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -739,49 +757,102 @@ class Deserializer {
         for (element_number_ = 0; element_number_ != msg_data.size; ++element_number_) {
-            void* element_pointer = component.advance_ptr(scenario_pointer, element_number_);
+            BufferView const element_buffer = advance(buffer, element_number_);
             // check the element is map or array
             auto const element_visitor = parse_map_array<visit_map_array_t, move_forward>();
             if (element_visitor.is_map) {
-                parse_map_element(element_pointer, element_visitor.size, component);
+                parse_map_element(element_buffer, element_visitor.size, component);
             } else {
-                parse_array_element(element_pointer, element_visitor.size, attributes);
+                parse_array_element(element_buffer, element_visitor.size, component, attributes);
             }
         }
         element_number_ = -1;
         offset_ = 0;
     }
 
-    void parse_map_element(void* element_pointer, Idx map_size, MetaComponent const& component) {
+    void parse_map_element(BufferView const& buffer, Idx map_size, MetaComponent const& component) {
         while (map_size-- != 0) {
             attribute_key_ = parse_string();
-            Idx const found_idx = component.find_attribute(attribute_key_);
-            if (found_idx < 0) {
+            Idx const component_attribute_idx = component.find_attribute(attribute_key_);
+            if (component_attribute_idx < 0) {
                 attribute_key_ = {};
                 // allow unknown key for additional user info
                 parse_skip();
                 continue;
             }
-            parse_attribute(element_pointer, component.attributes[found_idx]);
+            parse_attribute(buffer, component, component.attributes[component_attribute_idx]);
         }
         attribute_key_ = "";
     }
 
-    void parse_array_element(void* element_pointer, Idx array_size, std::span<MetaAttribute const* const> attributes) {
+    void parse_array_element(BufferView const& buffer, Idx array_size, MetaComponent const& component,
+                             std::span<MetaAttribute const* const> attributes) {
         if (array_size != static_cast<Idx>(attributes.size())) {
             throw SerializationError{
                 "An element of a list should have same length as the list of predefined attributes!\n"};
         }
+
         for (attribute_number_ = 0; attribute_number_ != array_size; ++attribute_number_) {
-            parse_attribute(element_pointer, *attributes[attribute_number_]);
+            parse_attribute(buffer, component, *attributes[attribute_number_]);
         }
         attribute_number_ = -1;
     }
 
-    void parse_attribute(void* element_pointer, MetaAttribute const& attribute) {
+    void parse_attribute(BufferView const& buffer, MetaComponent const& component, MetaAttribute const& attribute) {
+        assert(buffer.buffer != nullptr);
+
+        if (buffer.buffer->data != nullptr) {
```

Comment:
This is very tight loop to do duplicated checking on columnar data. We need to abstract this in the `parse_component` level. You can make `parse_scenario`, `parse_map_element` and `parse_array_element` by having a template argument to solve it compile time.

Reply:
If we're going to copy `AttributeBuffer`s, then this is no longer an expensive step. copying the buffers will be more expensive

Reply:
Copying `AttributeBuffer`s is like copying 640 bytes of data even with 20 columns. This tight loop check is about 1k scenario X 1k element X 20 attributes = 20 million checks per component.